### DISCUSSION TITLE
Date in Columns

### DISCUSSION THREAD

**Discussion by Zahar661** [serenity-is/Serenity#5614] [Help]
Hello everyone! 
Does anyone have any idea how this functionality can be implemented?
There are 2 tables, one with tasks, the second with execution, and here the dates need to be shown as columns so that you can edit.
A type of Gantt chart.
![Task](https://user-images.githubusercontent.com/26430136/113872676-c1713600-97bc-11eb-8021-1b99709d6866.JPG)

**Comment by LSNicholls** [comment]
I'm intrigued, but not understanding your data model.  Where is the date coming from?  It should be in Tasks Detail, right?  If so, are the allowable dates constrained in any way ?  In other words, suppose you have a year's worth of data.  How sparse are your measurements (how often do you measure percent)?  And how do you decide what period to show as columns, are there always the same number of columns?

**Comment by Zahar661** [comment]
@LSNicholls Data from tasks. The period is a week for editing. To display the quarter, half-year and year. The percentage is calculated on the fly for the entire task, now when the parent subtask is changed, the percentage is filled in. We have made this functionality, but only on hand-drawn bootstrap forms.

**Reply by LSNicholls** [reply]
>Data from tasks. 

@Zahar661, sorry I asked where the *date* was coming from, not the data.

So, do you mean that the "subtask" is the task detail entity and there is a change date on this entity?

>The period is a week for editing. To display the quarter, half-year and year. 

This is not what your columns show.  You show a total of 12 columns and it looks like daily data.

What I am driving at, here, is that I think you could get what you want with either a view or a stored procedure as the source, but to do it you'd have to know how many columns there should be and what they represented. 

This set of columns would have to be consistent if you were trying to show this using a Serenity set of Row objects -- although frankly I'm not sure why you would want to do that.  Let's say you really do want to do that:

* If you said you wanted to show "the last 12 days" or something like this, that would be easy; your view or stored procedure would return 12 columns plus your rollup percentage and the task name (and presumably task ID or something).
* It would be slightly harder, but not much harder, if your columns were something like "Every Friday during this quarter", as long as you created a result set with the maximum number of weeks in a quarter and left some empty when you were mid-way through the quarter (you could always hide the ones you didn't need, on display).
* If your answer is "we show all the days, and sometimes we show the quarter, sometimes the half-year, and sometimes the full year", that would be a lot of columns.  It makes more sense to say "we can show all the days for the past month, or a rollup of weeks for the past quarter, or a rollup by quarter for the past half-year or year".  That's definitely doable, but I would probably use three separate views or stored procedures as the source for each of those three things, and swap what I showed according to what the user selected.

But you would have an easier time with this using a real reporting facility; If I wanted a gantt chart or your exact report, I would probably embed a real report into the application.  In my case, probably SSRS, or PowerBI, or maybe an XSLT-and-XML result.  (If you think I am kidding about gantt charts in SSRS, [here's a walkthrough](https://spacefold.com/lisa/2017/07/29/walkthrough-sql-not-range-bars-is-the-secret-sauce-to-ssrs-gantt-charts/) I wrote up about it.)

I have looked a bit at Serenity's reporting facilities, too, and it doesn't seem hard to use, and maybe the columns could be dynamic -- I think it would work -- if you did that.  

So... just say what the column requirements are.
